hooks/boot/01-kafo-hook-extensions.rb

```diff
@@ -14,13 +15,20 @@ def new_install?
     !File.exist?(success_file)
   end
 
-  def ensure_package(package, state = 'installed')
-    unless ['installed', 'absent', 'latest'].include?(state)
-      fail_and_exit('Incorrect package state supplied to ensure_package')
-    end
+  def ensure_packages(packages, state = 'installed')
+    return if packages.empty?
 
+    code = "package { ['#{packages.join('\', \'')}']: ensure => #{state} }"
     bin_path = Kafo::PuppetCommand.search_puppet_path('puppet')
-    execute("#{bin_path} resource package #{package} ensure=#{state}")
+
+    logger.debug("Ensuring #{packages.join(', ')} to package state #{state}")
+    stdout_and_stderr, status = Open3.capture2e("echo \"#{code}\" | #{bin_path} apply --detailed-exitcodes")
+
+    unless [0, 2].include?(status.exitstatus)
+      log_and_say(:error, "Failed to ensure #{packages.join(', ')} are #{state}")
+      log_and_say(:error, stdout_and_stderr)
```

Comment:
Testing this, I don't think this information provides much value:

```
[ INFO 2020-02-25T13:29:10 verbose] Notice: Compiled catalog for puma.war.example.com in environment production in 0.63 seconds
Notice: Applied catalog in 0.87 seconds
```

Reply:
That's weird. Do you have more context because it should only be showing Puppet output when it fails but this looks like it did nothing at all.

Reply:
That output is actually form me printing it on successful install from testing to see if `stdout_and_stderr` provided any valuable logging output. When it does actually fail this line prints `nothing`:

```
Failed to ensure rh-postgresql12-postgresql-server, rh-redis5-redis are installed
[ERROR 2020-02-25T13:29:10 verbose] Failed to ensure rh-postgresql12-postgresql-server, rh-redis5-redis are installed
[ INFO 2020-02-25T13:29:10 verbose] Installer finished in 11.248545533 seconds
```

Reply:
It feels like the status.exitstatus doesn't match `0` or `2`. Could you add a `logger.debug("Exit status is #{status.exitstatus.inspect}")` as well?

Reply:
After playing around a bit, I found the output from switching to:

```
stdout, stderr, status = Open3.capture3

log_and_say(:error, stderr.strip)
```

To be cleaner. The inclusion of stdout just adds this kinda of crufty output:

```
Notice: Compiled catalog for puma.war.example.com in environment production in 0.70 seconds
```